fs/io_zenfs.cc

```diff
@@ -632,6 +633,171 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
+  std::vector<ZoneExtent*>::iterator e_it;
+  IOStatus s;
+  uint64_t r_pos;
+  uint32_t size;
+  uint32_t long_ext_size;
+  uint64_t new_start;
+  const char* ptr;
+  int dont_read = 0;
+
+  // Sort the Extent list in decreasing order.
+  std::sort(extent_list.begin(), extent_list.end(),
+            [](ZoneExtent* ext1, ZoneExtent* ext2) {
+              return ext1->length_ > ext2->length_;
+            });
+
+  // Get the size of the largest extent.
+  long_ext_size = extent_list[0]->length_;
+
+  // Allocate a buffer with size of the largest extent.
+  // We have to issue pread from the source zones so we
+  // need a buffer.
+  ptr = new char[long_ext_size];
+
+  zone_it = dst_zone_list.begin();
+  for (ext_it = extent_list.begin(); ext_it != extent_list.end();) {
+    ZoneExtent* ext;
+    Zone* zone_dst;
+
+    ext = *ext_it;
+    zone_dst = *zone_it;
+
+    // Set the position and length in the source zone to
+    // read the data.
+    r_pos = ext->start_;
+    size = ext->length_;
+    Slice buf(ptr, size);
+
+    if (!dont_read) s = ReadExtent(&buf, r_pos, ext->zone_);
+
+    // Free the allocated memory if error.
+    if (!s.ok()) {
+      if (ptr) delete[] ptr;
+      return s;
+    }
+
+    // Store the new starting position for the extent
+    // which will be later made persistent.
+    new_start = zone_dst->wp_;
+
+    // Write the valid data where were read from the
+    // source zone to the destination zone.
+    s = zone_dst->Append((char*)ptr, size);
+    if (s.ok()) {
+      // Data was written to the new zone, so the extent
+      // will have a new starting position. No need to
+      // change the length of the extent as it will be the
+      // same.
+      ext->start_ = new_start;
+
+      // The extent was moved to a new zone so change the
+      // resident zone parameter of the extent.
+      ext->zone_ = zone_dst;
+
+      // Current extent was written so now fetch the next extent.
+      ext_it++;
+      memset((char*)ptr, 0, long_ext_size);
+      continue;
+    }
+
+    if (s == IOStatus::NoSpace()) {
+      // Data was already read before, no need to read it again
+      dont_read = 1;
+
+      // The current zone cannot fit this extent because of lack
+      // of space, so get the next zone from the dst_zone_list.
+      zone_it++;
+    }
+
+    // There was an error so cannot proceed and simply
+    // we return the status.
+    if (s == IOStatus::IOError()) {
+      // If memory was allocated, we free it before returning.
+      if (ptr) delete[] ptr;
+
+      return s;
+    }
+  }
+
+  if (ptr) delete[] ptr;
+
+  return IOStatus::OK();
+}
+
+IOStatus ZenFSGCWorker::UpdateMetadataAfterMerge() {
+  std::vector<ZoneFile*>::iterator zone_file_it;
+  IOStatus s;
+  for (zone_file_it = files_moved_to_dst_zone.begin();
+       zone_file_it != files_moved_to_dst_zone.end(); zone_file_it++) {
+    ZoneFile* file_moved;
+    file_moved = *zone_file_it;
+
+    // What if the file is deleted before coming here?
```

Comment:
I believe we should lock the metadata and scan if there is any new records before updating metadata. This seems not to be easy.

Reply:
`IOStatus s;` is never assigned or modified throughout this function...

Reply:
> I believe we should lock the metadata and scan if there is any new records before updating metadata. This may require more careful design work.

 SyncFileMetadata() function calls the PersistRecord() function which holds lock before updating the metadata so I did not use any lock. Maybe you are right. I was thinking maybe we can remove the UpdateMetadataAfterMerge() completely and somehow figure out to update the metadata changes in the MoveValidDataToNewDestZone() function itself. It would have been less code and easier because why to have one more function.  But to do that we need to get the reference to the files whose extents have been moved. From the files, its easier to get the reference to the extents but not sure how to get the reference of the files from the extents.

Reply:
Looks good. If possible, please update the design doc for new information.